src/progress.rs

```diff
@@ -275,61 +450,167 @@ impl ProgressSet {
             if id == perspective_of {
-                active += 1;
+                active.insert(id);
                 continue;
             }
             if pr.recent_active {
-                active += 1;
+                active.insert(id);
             }
             pr.recent_active = false;
         }
         for (&_id, pr) in self.learners_mut() {
             pr.recent_active = false;
         }
-        active >= majority(self.voter_ids().len())
+        self.configuration.has_quorum(&active) &&
+            // If `next` is `None` we don't consider it, so just `true` it.
+            self.next_configuration.as_ref().map(|next| next.has_quorum(&active)).unwrap_or(true)
     }
 
     /// Determine if a quorum is formed from the given set of nodes.
+    ///
+    /// This is the only correct way to verify you have reached a quorum for the whole group.
+    #[inline]
     pub fn has_quorum(&self, potential_quorum: &FxHashSet<u64>) -> bool {
-        potential_quorum.len() >= majority(self.voter_ids().len())
+        self.configuration.has_quorum(potential_quorum) && self
+            .next_configuration
+            .as_ref()
+            .map(|next| next.has_quorum(potential_quorum))
+            // If `next` is `None` we don't consider it, so just `true` it.
+            .unwrap_or(true)
+    }
+
+    /// Determine if the ProgressSet is represented by a transition state under Joint Consensus.
+    #[inline]
+    pub fn is_in_membership_change(&self) -> bool {
+        self.next_configuration.is_some()
+    }
+
+    /// Enter a joint consensus state to transition to the specified configuration.
+    ///
+    /// The `next` provided should be derived from the `ConfChange` message. `progress` is used as
+    /// a basis for created peer `Progress` values. You are only expected to set `ins` from the
+    /// `raft.max_inflights` value.
+    ///
+    /// Once this state is entered the leader should replicate the `ConfChange` message. After the
+    /// majority of nodes, in both the current and the `next`, have committed the union state. At
+    /// this point the leader can call `finalize_config_transition` and replicate a message
+    /// commiting the change.
+    ///
+    /// Valid transitions:
+    /// * Non-existing -> Learner
+    /// * Non-existing -> Voter
+    /// * Learner -> Voter
+    /// * Learner -> Non-existing
+    /// * Voter -> Non-existing
+    ///
+    /// Errors:
+    /// * Voter -> Learner
+    /// * Member as voter and learner.
+    /// * Empty voter set.
+    pub fn begin_membership_change(
+        &mut self,
+        next: impl Into<Configuration>,
+        mut progress: Progress,
+    ) -> Result<(), Error> {
+        let next = next.into();
+        next.valid()?;
+        // Demotion check.
+        if let Some(&demoted) = self
+            .configuration
+            .voters
+            .intersection(&next.learners)
+            .next()
+        {
+            Err(Error::Exists(demoted, "learners"))?;
```

Comment:
No return? Amazing it can work.

Reply:
The `?` makes it return. Maybe I can make it more clear.